### DISCUSSION TITLE
`instanceof Tag` no longer works in version 0.3.0

### DISCUSSION THREAD

**Discussion by paulrudy** [markdoc/markdoc#386] [Bug report] [answered]
I have been using code like this:
```ts
if (children.length === 1 && children[0] instanceof Tag) {
  return children
}
```
as part of a custom paragraph node to identify image tags within paragraphs (as described in [this comment](https://github.com/markdoc/markdoc/issues/113#issuecomment-1201016102)).

Since upgrading to markdoc 0.3.0, the image tags no longer return `true` as `instanceof Tag`, although `child['$$mdtype'] === 'Tag')` _does_ return `true`.

I found issue #169, which seems to indicate that the `instanceof Tag` method should still work. Should it?

Or if this is an expected change with version 0.3.0, is `child['$$mdtype'] === 'Tag'` ok to use?

**Comment by mfix-stripe** [comment, marked as answer]
@paulrudy I recommend using `Tag.isTag(...)` to determine if an object is a tag or not. 

Can you drop a reproduction of your issue somewhere? I'd like to check out why this stopped working (or if it ever should have in the first place). Thanks!

(Cc @rpaul-stripe)

**Reply by matv-stripe** [reply]
Sometimes if you have a hot module reloading set up, webpack can load in different objects, so the `Tag` object that is imported could be different than the `Tag` object that `children[0]` is. I wonder if this error has a similar root cause?

**Reply by rpaul-stripe** [reply]
In addition to hot reloading, there are also a bunch of scenarios with module bundling that can result in the same class being imported multiple times from different bundles, which can lead to two instances that are ostensibly of the same class having different class identity from the perspective of the runtime. This phenomenon regularly break things like `instanceof` and equality checks on ES6 symbol values.

This is why it's useful to have the `$$mdtype` property with a literal value that will always match in an equality check. The `isTag` static method uses that property, so it's generally the safest way to check if a value is a `Tag` instance. I don't think anything in the 0.3.0 release changes the behavior of the `Tag` class in any way, so I would guess that there is an environmental factor that is in play. Either way, it's definitely better to use `isTag`, and that should solve the problem.

**Reply by paulrudy** [reply]
Thanks for your responses! @mfix-stripe I'll try to set up a Stackblitz reproduction for you soon

> In addition to hot reloading, there are also a bunch of scenarios with module bundling that can result in the same class being imported multiple times from different bundles, which can lead to two instances that are ostensibly of the same class having different class identity from the perspective of the runtime.

@rpaul-stripe I think that's exactly what's happening, if I understand you correctly. I'm using [Astro](https://github.com/withastro/astro) with its own markdoc integration, `@astrojs/markdoc`, and then also importing `{ Tag }` from `@markdoc/markdoc` to use in my custom paragraph node.

In any case, `Tag.isTag(...)` works, thanks for that :)

**Reply by paulrudy** [reply]
@mfix-stripe Ok, here's a [reproduction on Stackblitz](https://stackblitz.com/edit/github-g12f4c?file=src%2Fmarkdoc%2Fnodes%2Fparagraph.markdoc.js) using Astro showing how `instanceof Tag` stops working in v0.3.0.

And [reproducing](https://codesandbox.io/p/sandbox/relaxed-curie-yetpim?file=%2Fpages%2Fdocs%2Findex.md%3A13%2C328)  the same custom node using Next.js, there's no issue.

**Reply by mfix-stripe** [reply]
Yeah I imagine this is because `@astrojs/markdoc` uses a different version of `@markdoc/markdoc` (https://github.com/withastro/astro/blob/main/packages/integrations/markdoc/package.json#L42), and that creates conflicts when using instanceof

@rpaul-stripe we should maybe suggest to the Astro team to use peer dependencies for Markdoc. This is the same approach Apollo does with `graphql`: https://github.com/apollographql/apollo-client/blob/main/package.json#L67

**Reply by bholmesdev** [reply]
Ah, good point @mfix-stripe! We don't have clear instructions on importing Markdoc helpers without version conflicts. We expose the `Markdoc` default export from our own package, but that's probably not natural for lib maintainers. Can make this a peer dep
